src/librustc_traits/chalk/lowering.rs

```diff
@@ -554,41 +657,36 @@ impl<'tcx> LowerInto<'tcx, Option<chalk_ir::QuantifiedWhereClause<RustInterner<'
         self,
         interner: &RustInterner<'tcx>,
     ) -> Option<chalk_ir::QuantifiedWhereClause<RustInterner<'tcx>>> {
-        // FIXME(chalk): forall
-        match self.bound_atom(interner.tcx).skip_binder() {
+        let (predicate, binders, _named_regions) =
+            collect_bound_vars(interner, interner.tcx, &self.bound_atom(interner.tcx));
+        let value = match predicate {
             ty::PredicateAtom::Trait(predicate, _) => {
-                let predicate = ty::Binder::bind(predicate);
-                let (predicate, binders, _named_regions) =
-                    collect_bound_vars(interner, interner.tcx, &predicate);
-
-                Some(chalk_ir::Binders::new(
-                    binders,
-                    chalk_ir::WhereClause::Implemented(predicate.trait_ref.lower_into(interner)),
-                ))
+                Some(chalk_ir::WhereClause::Implemented(predicate.trait_ref.lower_into(interner)))
             }
             ty::PredicateAtom::RegionOutlives(predicate) => {
-                let predicate = ty::Binder::bind(predicate);
-                let (predicate, binders, _named_regions) =
-                    collect_bound_vars(interner, interner.tcx, &predicate);
-
-                Some(chalk_ir::Binders::new(
-                    binders,
-                    chalk_ir::WhereClause::LifetimeOutlives(chalk_ir::LifetimeOutlives {
-                        a: predicate.0.lower_into(interner),
-                        b: predicate.1.lower_into(interner),
-                    }),
-                ))
+                Some(chalk_ir::WhereClause::LifetimeOutlives(chalk_ir::LifetimeOutlives {
+                    a: predicate.0.lower_into(interner),
+                    b: predicate.1.lower_into(interner),
+                }))
+            }
+            ty::PredicateAtom::TypeOutlives(predicate) => {
+                Some(chalk_ir::WhereClause::TypeOutlives(chalk_ir::TypeOutlives {
+                    ty: predicate.0.lower_into(interner),
+                    lifetime: predicate.1.lower_into(interner),
+                }))
+            }
+            ty::PredicateAtom::Projection(predicate) => {
+                Some(chalk_ir::WhereClause::AliasEq(predicate.lower_into(interner)))
             }
-            ty::PredicateAtom::TypeOutlives(_predicate) => None,
-            ty::PredicateAtom::Projection(_predicate) => None,
             ty::PredicateAtom::WellFormed(_ty) => None,
 
             ty::PredicateAtom::ObjectSafe(..)
             | ty::PredicateAtom::ClosureKind(..)
             | ty::PredicateAtom::Subtype(..)
             | ty::PredicateAtom::ConstEvaluatable(..)
             | ty::PredicateAtom::ConstEquate(..) => bug!("unexpected predicate {}", &self),
-        }
+        };
+        value.map(|value| chalk_ir::Binders::new(binders, value))
```

Comment:
This can be simplified by making `value` a `WhereClause` instead of `Option<WhereClause>`, and doing `return None` for the WellFormed case.